packages/upward-js/lib/ResolverVisitor.js

```diff
@@ -0,0 +1,107 @@
+const debug = require('debug')('upward-js:ResolverVisitor');
+const { inspect } = require('util');
+const { ResolverList, ResolversByType } = require('./resolvers');
+const { zipObject } = require('lodash');
+const isPrimitive = require('./isPrimitive');
+
+class ResolverVisitor {
+    constructor(io, rootDefinition, context) {
+        this.io = io;
+        this.rootDefinition = rootDefinition;
+        this.context = context;
+        this.context.setVisitor(this);
+    }
+    async downward(contextNames) {
+        debug('resolving downward: %o', contextNames);
+        let passedMiddleware = false;
+        const valuePromises = contextNames.map(async name => {
+            const value = await this.upward(this.rootDefinition, name);
+            if (typeof value === 'function') {
+                debug(
+                    '%s request returned a function, we are assuming it is a middleware'
+                );
+                passedMiddleware = value;
+                throw new Error('PASSED_MIDDLEWARE');
+            }
+            return value;
+        });
+        try {
+            const values = await Promise.all(valuePromises);
+            return zipObject(contextNames, values);
+        } catch (e) {
+            if (e.message === 'PASSED_MIDDLEWARE') {
+                debug(
+                    `returning middleware from visitor.downward() instead of object`
+                );
+                return passedMiddleware;
+            } else {
+                throw e;
+            }
+        }
+    }
+    async upward(definition, propertyName) {
+        debug('resolving upward: %s from %o', propertyName, definition);
+        if (!definition.hasOwnProperty(propertyName)) {
+            throw new Error(
+                `Context value '${propertyName}' not defined in ${inspect(
+                    definition
+                )}.`
+            );
+        }
+        const defined = definition[propertyName];
+
+        const resolver = this.getResolverFor(defined, propertyName);
+
+        if (resolver) {
+            return resolver.resolve(defined);
+        }
+
+        if (isPrimitive(defined)) {
+            const definedString = defined.toString();
+            debug(
+                'defined: %s is primitive, yielding to context.get("%s")',
+                definedString,
+                definedString
+            );
+            return this.context.get(definedString);
+        }
+
+        if (typeof defined !== 'object' || !this.getResolverFailure) {
+            throw new Error(`Unexpected value in config: ${defined}`);
+        } else {
+            throw new Error(this.getResolverFailure);
+        }
+    }
+    getResolverFor(defined, propertyName) {
+        let Resolver;
+        for (Resolver of ResolverList) {
```

Comment:
The scoping of `Resolver` is misleading here.

- `Resolver` is declared _as a mutable variable_ prior to the `for...of` block
- On each loop iteration, `Resolver` is reassigned and `Resolver.recognize()` is always called
- On each loop iteration, there may be an early return

The only time this arrangement is useful is if the code following the `for...of` block is interested in the value of `Resolver` _after the loop has completed_. Otherwise, we would expect to see `for (const Resolver of ResolverList) {}`.

However, the code following the `for...of` block ("subsequent code") is not interested in the value of `Resolver`. In fact, the subsequent code always reassigns `Resolver` before ever reading its value. This is counterintuitive because `Resolver` appears to have been specifically declared such that it is meaningful, yet it isn't.

My recommendation:

```es6
getResolverFor(defined, propertyName) {
    for (const Resolver of ResolverList) {
        // ...
    }

    const definedResolver = defined.resolver

    const Resolver = definedResolver
        ? ResolversByType[definedResolver]
        : ResolverList.find(({ telltale }) => defined.hasOwnProperty(telltale))

    if (Resolver) {
        return new Resolver(this)
    } else {
        this.getResolverFailure = definedResolver
            ? 'Unrecognized resolver type...'
            : 'Unrecognized configuration...'
    }
}

```